### DISCUSSION TITLE
How i can call an api from /api folder in server component without known the domain name before the time?

### DISCUSSION THREAD

**Discussion by mandorakannu** [vercel/next.js#49740] [Help] [answered]
### Summary

Hi, I am trying to fetch data from an api from /api folder without known the domain name before the time.
I mean i am try to create a static pages of blog which is fetch on build time but on vercel, how can i know the domain name before the build. If i don't know how the data will going to fetch???

Please explain to me it is very important question for me..

Thank You❤️

**Comment by schrodienieur** [comment]
Have you tried using relative URL to fetch the data? Like fetch(/api/posts).

If you deploy it in Vercel. You can use the system environment variables to know the domain name. 

https://vercel.com/docs/concepts/projects/environment-variables/system-environment-variables

There is variable named VERCEL_URL, it is the domain name of the generated deployment URL. Example: *.vercel.app. The value does not include the protocol scheme https://.

**Reply by mandorakannu** [reply]
Can you please explain in step? Its hard to understand for me

**Reply by schrodienieur** [reply, marked as answer]
You can fetch the data by using VERCEL_URL environment variable to make the absolute URL.
For example, 


await fetch("https://" + process.env.VERCEL_URL + "/api/post")

**Reply by mandorakannu** [reply]
So, VERCEL_URL is automatically assign right?

**Reply by schrodienieur** [reply]
Yes, it's already in the system environment.

**Reply by mandorakannu** [reply]
Thank you very much, Sir! I appreciate you❤️

**Reply by mandorakannu** [reply]
> You can fetch the data by using VERCEL_URL environment variable to make the absolute URL. For example,
> 
> await fetch("https://" + process.env.VERCEL_URL + "/api/post")

I try this and i got the domain name but there is an issue. it doesn't fetch the data from the api and crash the production build
with this error: `Unexpected token < in JSON at position 0`

**Reply by fuma-nama** [reply]
You can't do that because API Routes aren't running during build time. You have to consider about alternatives like directly interacting with database (in server components)

**Reply by cryptickoan** [reply]
calling the api from a server component will not work with SSG, only with SSR. 

It can be helpful because with SSR there's no way to cache db calls, or any other type of call that isnt using fetch(), so by fetching the api you can rely on the route handler cache. 

Downside is that it does trigger a new request from your server component to your api since these are not bridged internally

**Comment by fuma-nama** [comment]
As your pages are statically generated, Next.js will run the `fetch` function in order to generate static HTML code. 

However, Route Handlers won't be running during the build time. 
The fetch will fail with a 404 error because the API is not currently available, you must disable static generation or run the code in a server component directly

For example, change this: 
```tsx
export default function BlogPage() {
    const blogs = await fetch(VERCEL_URL + "/api/blog", { ... });

    return <div>{...}</div>
}
```
to this:
```tsx
async function getBlogPosts() {
    //fetch blogs from remote sources or database

    return prisma.findMany({ ... })
}

export default function BlogPage() {
    const blogs = await getBlogPosts()

    return <div>{...}</div>
}
```
